[Low Zhen Dong Leon] iP

leonlowzd opened the pull request.

leonlowzd committed:
Level 0. Greet

Implement an initial skeletal version of the Duke that simply greets the user and exits.

leonlowzd committed:
Level 1. Greet, Echo, Exit

Implement a skeletal version of Duke that starts by greeting the user, simply echos commands entered by the user, and exits when the user types bye.

Added another feature to remove spaces if user inputs "bye " or " bye ".

leonlowzd committed:
Level 2. Add, List

Add the ability to store whatever text entered by the user and display them back to the user when requested.

leonlowzd committed:
Level 3. Mark as Done

Add the ability to mark tasks as done.

leonlowzd committed:
A-CodingStandard

Tweak the code to comply with a coding standard.

leonlowzd committed:
Level-4: Work In Progress
Major changes, code is more OOP now. Still have to review and test code

leonlowzd committed:
Task 4: Cleanup plus event bug fixes, almost ready to tag and submit

leonlowzd committed:
Level 4: ToDos, Events, Deadlines

Added 3 types of Tasks and made Duke more robust: if user inputs something unrecognizable, programme will not crash

leonlowzd committed:
A-CodeQuality: Improve Code Quality

1. Cleans up the code overall
2. In the main loop, nested if-else condition have been replaced by a switch case.
3. Improved extraction of user input string
4. Better commenting of code

tobiasceg commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,147 @@
+            String printStatement = null;
+
+            switch (operation){
+                case "bye":
```
Switch statements: as per the coding standards, switch statements do no require indentations

tobiasceg replied:
Switch statements: as per the coding standards, switch statements do no require indentations

tobiasceg commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,147 @@
-    }
-}
+
+        boolean exitFlag = false;
```
boolean naming could be written to sound like a boolean : hasExit()

tobiasceg reviewed (commented):
tutorial PR 1

prachi2023 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,147 @@
-}
+
+        boolean exitFlag = false;
+        int index = 0;
```
I feel like this could be a more descriptive name to show what it is indexing.

prachi2023 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,147 @@
+    }
+
+    // This function extracts the operation type from the user input's String
+    private static String extractOperationType(String userInput){
+        userInput = userInput.trim();
+        String [] operation = userInput.split(" ");
+        return operation[0];
+
+    }
+
+    // This function removes the task type and extracts the description of the task
+    private static String extractDescriptionFromString(String type, String userInput){
+        String description;
+        if (type.equals("todo")){
+            description = userInput.replace(type, "");
+        } else{
+            description = userInput.replace(type, "");
+            description = description.substring(0, description.lastIndexOf("/"));
+        }
+        return description;
+    }
```
I like the way you have named these methods, makes easy to tell what they do

tobiasceg commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,147 @@
+    }
+
+    // This function constructs the printout of the newly added task
+    private static String printTaskDescription(int index,  Task[] list){
+        int numberOfTask = index+1;
+        String printStatement = "____________________________________________________________\n"
+                + "Got it. I've added this task:\n"
+                + "     " + list[index] + "\n"
+                + "Now you have "+numberOfTask+" in the list.\n"
+                + "____________________________________________________________\n";
+        return printStatement;
```
Great use of Refractors, this improves readability

tobiasceg commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,147 @@
+            String printStatement = null;
+
+            switch (operation){
+                case "bye":
+                    exitFlag = true;
+                    printStatement = printByeMessage();
+                    break;
+
+                case "list":
+                    printStatement = printFullList(list, index);
+                    index--;
+                    break;
+
+                case "done":
+                    markTaskAsDone(list,line);
+                    index--;
+                    break;
+
+                case "todo":
+                    String toDoDescription = extractDescriptionFromString("todo",line);
+                    ToDo t = new ToDo(toDoDescription);
+                    list[index] = t;
+                    printStatement = printTaskDescription(index, list);
+                    break;
+
+                case "deadline":
+                    String deadlineDescription = extractDescriptionFromString("deadline", line);
+                    String date = extractDateFromString(line);
+                    Deadline d = new Deadline(deadlineDescription, date);
+                    list[index] = d;
+                    printStatement = printTaskDescription(index, list);
+                    break;
+
+                case "event":
+                    String eventDescription = extractDescriptionFromString("event", line);
+                    String eventDate = extractDateFromString(line);
+                    Event e = new Event(eventDescription, eventDate);
+                    list[index] = e;
+                    printStatement = printTaskDescription(index, list);
+                    break;
```
Great use of switch statement! Really improves readability and produces a clean looking code

tobiasceg commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,147 @@
+    // This function prints the full list of Tasks
+    private static String printFullList(Task[] list, int index) {
+        String printStatement;
+        // This object acts as a buffer to build strings: they are based on mutable character arrays
+        // This to reduce the cost of growing the string
```
Great comments to help other coders to understand the purpose of the code

leonlowzd committed:
Level 5: Handle Errors

1. handle most of the errors such as invalid done indexes, invalid dates, invalid operations, empty descriptions'

2. Improved variable names and code flow

3. Created new objects for exceptions

leonlowzd committed:
Level 5.1: Handle Errors- improved version plus minor name changes

1. handle most of the errors such as invalid done indexes, invalid dates, invalid operations, empty descriptions'

2. Improved variable names and code flow

3. Created new objects for exceptions

leonlowzd committed:
Level 5.2: Handle Errors- improved version plus minor name changes

1. handle most of the errors such as invalid done indexes, invalid dates, invalid operations, empty descriptions'

2. Improved variable names and code flow

3. Created new objects for exceptions

leonlowzd committed:
Removed unnecessary comments

leonlowzd committed:
Revert "Removed unnecessary comments"

This reverts commit a259c5c2bd95e222cc497c87e9221797a8e989e0.

leonlowzd committed:
Organised code into packages

leonlowzd committed:
Level 7. Save

Save the tasks in the hard disk automatically whenever the task list changes. Load the data from the hard disk when Duke starts up. You may hard-code the file name and location e.g., [project_root]/data/duke.txt

leonlowzd committed:
Level 6. Delete

Add support for deleting tasks from the list.

leonlowzd committed:
Merge tag 'Level-6'

leonlowzd committed:
Resolved merge conflict with Level-7 and minor exception recovery

leonlowzd committed:
Level-7: added new class
1.Enabled both read and write

leonlowzd committed:
Level-7: added new class
1.Enabled both read and write

**Merge Conflict resolved with master branch

leonlowzd committed:
Level-7: added new class
1.Enabled both read and write

Major release, left reading date to be fixed

leonlowzd committed:
Level-7: added new class
1.Enabled both read and write

Major release, left reading date to be fixed

leonlowzd committed:
Level-7: Fixed the date issue, close to finalising Level-7

leonlowzd committed:
Level-7: Fixed error when Duke tries to read the file that doesn't exist at the given directory, The program will just return empty to continue normal operations

leonlowzd committed:
Level-7: All features tested, fully working, final version for tmr's demo

leonlowzd committed:
Level-7: Completed
Changed the txt format to a easier to manipulate format. Tested on both Windows and Mac

JustinnT commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,253 @@
+        boolean hasExit = false;
+
+        // Problem will come if there are more than 100 tasks
+//        Task[] list = new Task[100];
```
Should line 22 comment be there?

leonlowzd replied:
Should be removed... old code

JustinnT commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,253 @@
+                case "deadline":
+                case "event":
+                    try {
+                        if(operation.equals("todo")) printStatement = createNewTask(operation,extractDescriptionFromString(operation,line),null,false);
```
For this if else statement, should it be covered by curly brackets and have the Egyptian style? Also is this line too long and should there be a whitespace after commas?

leonlowzd replied:
Yes should have curly brackets with Egyptian styling. 

Yes, lines are too long, shall break it down with the proper whitespace commas

JustinnT commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,253 @@
+
+            }
+            System.out.println(printStatement);
+            save.writeFile(home+"/Documents/log.txt",printFullList(numberOfTasks));
```
Same here as well, should there be a white space before and after the "+" operator and commas?

leonlowzd replied:
yes 👍

JustinnT commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,253 @@
+        }
+    }
+    
+    public static String createNewTask(String taskType, String description, String date, Boolean status) throws IllegalDate, IllegalDescription {
```
Is this line too long?

leonlowzd replied:
too long, will break it down

JustinnT commented on src/main/java/Duke/Deadline.java:
```diff
@@ -0,0 +1,34 @@
+    public Deadline(String description, String date) throws IllegalDate {
+        super(description);
+        if(date.contains("/by")) {
+            if (date.replace("/by","").trim().isEmpty()) throw new IllegalDate();
```
Should the if statement be 1 line? Also this might be a potential arrowhead code. What are the ways you can avoid arrowhead?

JustinnT commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,253 @@
+            String printStatement;
+
+            switch (operation){
+                case "bye":
```
I think it would be better to express the case conditions like "list" "bye" "done" etc. as constants.

leonlowzd replied:
planning to do so in the next commit..

JustinnT commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,253 @@
+                case "done":
+                    try {
+                        printStatement = markTaskAsDone(line);
+                    } catch ( IllegalIndex e) {
```
Should there be a whitespace after the bracket?

JustinnT commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,253 @@
+                break;
+
+            case "event":
+                Event e = new Event(description, date);
```
Perhaps can use a more descriptive variable name here.

leonlowzd replied:
So something like "Event event = new Event(description, date)"?

JustinnT replied:
Yup that can something like that. I was just concerned about e being too vague because while the context is different, "e" was used in one of the catch statement (catch (exception e)). So its good to just specify the variable names👍

JustinnT reviewed (commented):
Good job but have to check if your lines exceed 120 characters, also, I noticed that there were some 1 line if statements.

leonlowzd committed:
Level 7: bug fixes & code cleanup
 fixed the error when the index of the list gets out of bound due to incrementation of index when an exception is thrown. This is not required as decrement is done after the potential error catching line

leonlowzd committed:
Week-7 WIP

leonlowzd committed:
New master commit

leonlowzd committed:
Created TaskList class to store tasks, WIP

leonlowzd committed:
A-MoreOOP: WIP

leonlowzd committed:
A-MoreOOP: WIP, repacked some of the files.

leonlowzd committed:
A-MoreOOP: WIP, removed unnecessary try and catch phrases

leonlowzd committed:
A-MoreOOP: Finished converting to more OOP

leonlowzd committed:
Level-9: Added Find command

leonlowzd committed:
Level-9: Added Find command, fixed the escape key bug

leonlowzd committed:
Merge pull request #1 from leonlowzd/Level-9

Level-9: Merge with Master

leonlowzd committed:
Added JavaDoc for some of the public methods, to be finalised

leonlowzd committed:
A-JavaDoc: Add JavaDoc comments to the code.

leonlowzd committed:
Merge pull request #2 from leonlowzd/A-JavaDoc

A java doc

leonlowzd committed:
A-UserGuide: Added User Guide to the project

leonlowzd committed:
Set theme jekyll-theme-slate

leonlowzd committed:
Set theme jekyll-theme-slate

leonlowzd committed:
Update README.md

leonlowzd committed:
Update README.md

Minor edits

leonlowzd committed:
Updated README

leonlowzd committed:
Updated README.md

leonlowzd committed:
Level-9: added auto conversion to date format.

leonlowzd committed:
Added new exceptions into reading text files, able to hand exceptions and file corruptions

leonlowzd committed:
Merge pull request #3 from leonlowzd/heads/Level-8

Level-9: added auto conversion to date format.

leonlowzd committed:
Updated README.md

leonlowzd committed:
Update README.md

leonlowzd committed:
Update README.md

leonlowzd committed:
Update README.md

leonlowzd committed:
Updated README.md

leonlowzd committed:
Better error message printouts for handling dates

leonlowzd committed:
Update README.md

leonlowzd committed:
Improved comments for the entire program
